### ISSUE TITLE
DB deadlocks during migration

### ISSUE THREAD

**Issue by matty-v** [hasura/graphql-engine#7455]
We are attempting to upgrade our hasura engine from v1 to v2 - but we are noticing timeouts and deadlocks when applying migrations in our prod environment. The statement that hangs the DB is the one that drops all the tables storing hasura metadata:
```
-- This migration enables metadata separation
-- Drops all tables which stores hasura metadata
DROP VIEW hdb_catalog.hdb_role;
DROP TABLE hdb_catalog.hdb_custom_types;
DROP TABLE hdb_catalog.hdb_action_permission
...
```
The timeouts/deadlocks aren't an issue in our dev and test environments; we think that it's our normal prod traffic that is causing the hangups. Do you have any recommendations for how we can avoid these issues without needing a planned outage? And a followup question - once we apply the migration that upgrades to hasura v2, will future migrations need to drop and re-create all the hasura metadata tables? Or is it a one-time thing?

Thanks in advance.

**Comment by coco98** [comment]
@tirumaraiselvan what resolution would you recommend here? This might be because of the event trigger cleanup that's required before making the upgrade?

**Comment by tirumaraiselvan** [comment]
@matty-v What is the size (num of rows) in your hdb_catalog.event_log table? In 2.0, we add an index to this table and it could be a costly operation if the event_log table is huge. We recommend pruning this table to remove historical events before doing the upgrade: https://hasura.io/docs/latest/graphql/core/event-triggers/clean-up.html

**Comment by matty-v** [comment]
@tirumaraiselvan we've been following the recommendations in this doc: https://hasura.io/docs/latest/graphql/core/guides/upgrade-hasura-v2.html

> In case you happen to have a large number of past cron and event trigger logs in your database, this might slow down the update to v2 and might even cause unexpected errors in certain scenarios.
It is highly recommended to clean up past cron and event trigger logs data from the database before attempting the update.

We deleted all rows from `hdb_catalog.event_log` and `hdb_catalog.event_invocation_logs` before attempting the upgrade, but still ran into the deadlocks

**Comment by tirumaraiselvan** [comment]
@zee-hussain @matty-v Unfortunately, this is getting pretty hard to replicate on our end. Would it be possible to "clone" your prod environment and try it out there? The database locks query that I shared earlier will reveal what is blocking that query.

**Comment by rakeshkky** [comment]
@zee-hussain @matty-v Also, kindly give us following details regarding your production environment.
1. How many server instances are running on the database?
2. If there are multiple instances how are you performing the v2 upgrade? One instance at a time or all instances at once?

> once we apply the migration that upgrades to hasura v2, will future migrations need to drop and re-create all the hasura metadata tables? Or is it a one-time thing?

The Hasura metadata tables are dropped once for all while migrating to v2. We only rely on `hdb_catalog.hdb_metadata` table for metadata changes.

**Comment by ifokeev** [comment]
Same problem here during the upgrade, locks and deadlocks destroying the configuration

**Comment by ifokeev** [comment]
So I assume the problem is in two tables: `db_catalog.event_invocation_logs` and `hdb_catalog.event_log`. These tables have so many rows so that's causes long transactions which becoming stucked somehow. In production it means about 1-2 hours of outage in debugging it

**Comment by tirumaraiselvan** [comment]
@ifokeev That could be a likely cause since we create an index on the event_log table in the v1.3 to v2.0 migration. We recommend cleaning up historical events before upgrading: https://hasura.io/docs/latest/graphql/core/event-triggers/clean-up.html

**Comment by ifokeev** [comment]
@tirumaraiselvan could you add to the migrations a transaction and `TRUNCATE` before the `DROP`, because now the migration process is breaking user's production environment, I'm not the only one with this case

**Comment by tirumaraiselvan** [comment]
@ifokeev Can you give us the count of the `hdb_catalog.event_log` table

**Comment by ifokeev** [comment]
@tirumaraiselvan I already truncated the tables, sorry. Truncate works much faster than `DROP`, `hasura` has to fix it

**Comment by tirumaraiselvan** [comment]
@ifokeev Hasura doesn't truncate or drop the `hdb_catalog.event_log` table because it contains your historical events. We are figuring out a way to make the size of this table NOT affect the upgrade process.

**Comment by matty-v** [comment]
> @zee-hussain @matty-v Unfortunately, this is getting pretty hard to replicate on our end. Would it be possible to "clone" your prod environment and try it out there? The database locks query that I shared earlier will reveal what is blocking that query.

@tirumaraiselvan We have already tested the upgrade in one of our staging environments which is a clone of prod and did not see any issues. Our staging env uses a snapshot of the prod DB, so there aren't any differences in data. It makes me wonder if the deadlocks are related to the level of user traffic in prod, since our staging env is only used by us internally.

**Comment by matty-v** [comment]
> @zee-hussain @matty-v Also, kindly give us following details regarding your production environment.
> 
> 1. How many server instances are running on the database?
> 2. If there are multiple instances how are you performing the v2 upgrade? One instance at a time or all instances at once?

We are only running a single instance of Hasura

**Comment by rakeshkky** [comment]
@matty-v 
Thank you for your response.
>  It makes me wonder if the deadlocks are related to the level of user traffic in prod, since our staging env is only used by us internally.

As you said, you are only having one instance of Hasura GraphQL Engine on the database. If it is migrating, I don't think it'll serve any API in the production. We need to know your migration process briefly.

**Comment by matty-v** [comment]
@rakeshkky at a high level our migration process is to spin up an ephemeral docker container that starts the Hasura Engine and then runs a script on that container which executes `hasura migrate apply`. Once the migration finishes the container will tear down. 
This migration container will connect to the same database that is handling our normal prod traffic, and runs alongside our "live" instance of Hasura (which is running in its own container)

**Comment by rakeshkky** [comment]
@matty-v Thank you for your response
> ... that starts the Hasura Engine and then runs a script on that container which executes ...

May I know the version of "Hasura Engine" that is being run to execute `hasura migrate apply` ?

**Comment by matty-v** [comment]
@rakeshkky we were attempting to upgrade from v1.3.2 to v2.0.4.
So I think we would have been on v2.0.4 when running the `hasura migrate apply` command. And it was successful when running it in our test environments. But that was the command that ended up hanging and causing the deadlock errors when we tried it on our prod instance.

**Comment by tirumaraiselvan** [comment]
@matty-v Sorry for the delay in getting to the bottom of this. IIUC, you have a container which is serving live traffic. Then you launch a temp container of newer version which upgrades Hasura. But isn't the traffic container still on the older version? 

If that is the case, then are there metadata operations, etc happening in your traffic? This could delay the execution of the migration as it will be queued behind any DML operations on the metadata. For reference, this is the migration that you quoted in the first comment: https://github.com/hasura/graphql-engine/blob/master/server/src-rsr/migrations/42_to_43.sql

**Comment by matty-v** [comment]
@tirumaraiselvan that's correct - we would first be running the temp container to upgrade Hasura and apply migrations while live traffic is being handled by the container running the older version of Hasura. And as you mentioned, this is when we see the deadlocks that reference the sql you linked to. These deadlocks end up halting our prod traffic as well as cause our Hasura migration/upgrade to fail. 

Circling back to my original question, are there any suggestions on the best way to upgrade Hasura against our prod database without risking the deadlocks? Or is this a case where we need to have a planned outage and halt all live traffic?

Also, I'm assuming that https://github.com/hasura/graphql-engine/blob/master/server/src-rsr/migrations/42_to_43.sql only needs to be run once when we upgrade to Hasura engine v2, right? We don't want to have planned outages every time we run the `hasura migrate apply` command (which is part of our CI/CD pipeline).

Thanks!

**Comment by rakeshkky** [comment]
@matty-v 
> I'm assuming that https://github.com/hasura/graphql-engine/blob/master/server/src-rsr/migrations/42_to_43.sql only needs to be run once when we upgrade to Hasura engine v2, right?

Yes. It is the migration needed on the database (provided via `--database-url` or `HASURA_GRAPHQL_DATABASE_URL`) to serve v2.x GraphQL Engine. It only alters the schema present in `hdb_catalog` and drops entirely unused `hdb_views` schema. 

> We don't want to have planned outages every time we run the hasura migrate apply command

The `hasura migrate apply` command interacts with [run_sql](https://hasura.io/docs/latest/graphql/core/api-reference/schema-api/run-sql.html#run-sql) API. It should not cause any deadlocks. 

> are there any suggestions on the best way to upgrade Hasura against our prod database without risking the deadlocks? Or is this a case where we need to have a planned outage and halt all live traffic?

We need to know what your typical `live traffic` consists of.

**Comment by cep21** [comment]
We also ran into the same issue upgrading 1.3 -> 2.0

It locked the database, causing hasura instances to fail health checks in kubernetes, causing all hasura instances to go down.  This locking continued for around 15-20 minutes.  Here are some graphs we were able to capture during the event

This is AWS performance insights graph sorted by top SQL AAS

![image](https://user-images.githubusercontent.com/20358/136282162-bbc298cc-250b-4b7c-9812-f33732817980.png)

**Comment by tirumaraiselvan** [comment]
Closing this issue since #7711 is fixed and will be out in v2.1.0.
